lxd/util/resources.go

```diff
@@ -0,0 +1,337 @@
+				return nil, err
+			}
+
+			t = &thread{}
+			t.ID = uint64(id)
+
+			path := fmt.Sprintf("/sys/devices/system/cpu/cpu%d/topology/core_id", t.ID)
+			coreID, err := parseNumberFromFile(path)
+			if err != nil {
+				return nil, err
+			}
+
+			t.coreID = uint64(coreID)
+
+			path = fmt.Sprintf("/sys/devices/system/cpu/cpu%d/topology/physical_package_id", t.ID)
+			sockID, err := parseNumberFromFile(path)
+			if err != nil {
+				return nil, err
+			}
+
+			t.socketID = uint64(sockID)
+
+			path = fmt.Sprintf("/sys/devices/system/cpu/cpu%d/cpufreq/scaling_cur_freq", t.ID)
+			freq, err := parseNumberFromFile(path)
+			if err != nil {
+				return nil, err
+			}
+
+			t.frequency = uint64(freq / 1000)
+
+			path = fmt.Sprintf("/sys/devices/system/cpu/cpu%d/cpufreq/cpuinfo_max_freq", t.ID)
+			freq, err = parseNumberFromFile(path)
+			if err != nil {
+				return nil, err
+			}
+
+			t.frequencyTurbo = uint64(freq / 1000)
+
+			threads = append(threads, *t)
+		} else if strings.HasPrefix(line, "vendor_id") {
+			i := strings.Index(line, ":")
+			if i < 0 {
+				return nil, err
+			}
+			i++
+
+			line = line[i:]
+			line = strings.TrimSpace(line)
+
+			if t != nil {
+				threads[len(threads)-1].name = line
+			}
+		} else if strings.HasPrefix(line, "model name") {
+			i := strings.Index(line, ":")
+			if i < 0 {
+				return nil, err
+			}
+			i++
+
+			line = line[i:]
+			line = strings.TrimSpace(line)
+
+			if t != nil {
+				threads[len(threads)-1].vendor = line
+			}
+		}
+	}
+
+	if len(threads) == 0 {
+		return nil, db.NoSuchObjectError
+	}
+
+	return threads, err
+}
+
+func parseSysDevSystemCpu() ([]thread, error) {
+	ents, err := ioutil.ReadDir("/sys/devices/system/cpu/")
+	if err != nil {
+		return nil, err
+	}
+
+	threads := []thread{}
+	for _, ent := range ents {
+		entName := ent.Name()
+		if !strings.HasPrefix(entName, "cpu") {
+			continue
+		}
+
+		entName = entName[len("cpu"):]
+		idx, err := strconv.Atoi(entName)
+		if err != nil {
+			continue
+		}
+
+		t := thread{}
+		t.ID = uint64(idx)
+		path := fmt.Sprintf("/sys/devices/system/cpu/cpu%d/topology/core_id", t.ID)
+		coreID, err := parseNumberFromFile(path)
+		if err != nil {
+			return nil, err
+		}
+
+		t.coreID = uint64(coreID)
+
+		path = fmt.Sprintf("/sys/devices/system/cpu/cpu%d/topology/physical_package_id", t.ID)
+		sockID, err := parseNumberFromFile(path)
+		if err != nil {
+			return nil, err
+		}
+
+		t.socketID = uint64(sockID)
+
+		path = fmt.Sprintf("/sys/devices/system/cpu/cpu%d/cpufreq/scaling_cur_freq", t.ID)
+		freq, err := parseNumberFromFile(path)
+		if err != nil {
+			return nil, err
+		}
+
+		t.frequency = uint64(freq / 1000)
+
+		path = fmt.Sprintf("/sys/devices/system/cpu/cpu%d/cpufreq/cpuinfo_max_freq", t.ID)
+		freq, err = parseNumberFromFile(path)
+		if err != nil {
+			return nil, err
+		}
+
+		t.frequencyTurbo = uint64(freq / 1000)
+
+		threads = append(threads, t)
+	}
+
+	if len(threads) == 0 {
+		return nil, db.NoSuchObjectError
```

Comment:
That's an odd error to return for something that's not db based.

os.ErrNotExist would probably be a better fit here.

Reply:
This error previously didn't live in the db code. It has been moved there while free was redoing a bunch of stuff and so there are actually quite a lot of places in some of the storage codepaths where this needs to be replaced as well.

Reply:
Yeah, lets just make sure we don't add additional ones :)

Reply:
done